source/ring_buffer.c

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/ring_buffer.h>
+
+int aws_ring_buffer_init(struct aws_ring_buffer *ring_buf, struct aws_allocator *allocator, size_t size) {
+    AWS_ZERO_STRUCT(*ring_buf);
+
+    ring_buf->allocation = aws_mem_acquire(allocator, size);
+
+    if (!ring_buf->allocation) {
+        return AWS_OP_ERR;
+    }
+
+    ring_buf->allocator = allocator;
+    aws_atomic_store_ptr(&ring_buf->head, ring_buf->allocation);
+    aws_atomic_store_ptr(&ring_buf->tail, ring_buf->allocation);
+    ring_buf->allocation_end = ring_buf->allocation + size + 1;
+
+    return AWS_OP_SUCCESS;
+}
+
+void aws_ring_buffer_clean_up(struct aws_ring_buffer *ring_buf) {
+    if (ring_buf->allocation) {
+        aws_mem_release(ring_buf->allocator, ring_buf->allocation);
+    }
+
+    AWS_ZERO_STRUCT(*ring_buf);
+}
+
+int aws_ring_buffer_acquire(struct aws_ring_buffer *ring_buf, size_t requested_size, struct aws_byte_buf *dest) {
```

Comment:
Since ```acquire``` and ```acquire_up_to``` are basically the same except for the last section, why not just make one function that takes a ```size_t *returned_size```. If ```returned_size``` is non-null, do the best effort allocation. Then implement both of these in terms of that one function, less copy-pasta, only have to get it right once.

Reply:
Nevermind, they're not as identical as I thought.

Reply:
because the code is complicated enough as it is, and I didn't want to make it worse by framenting it across functions. There's subtle differences now, and it would be too messy to refactor it just to share less than 20 logical lines of code.